---
title: "`r paste('Analysis:', ifelse(grepl('bam', param$name, ignore.case = TRUE), sub('$', '_Count_QC', param$name), param$name))`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup countQC, include=FALSE}
#title: "`r paste('Analysis:', ifelse(grepl('bam', param$name, ignore.case = TRUE), sub('$', '_Count_QC', param$name), param$name))`"
# title: "debug title"

## This report requires the object rawData
require(knitr)
require(kableExtra)
require(SummarizedExperiment)
require(wesanderson)
require(WGCNA)
require(plotly)
require(matrixStats)
require(reshape2)
```

```{r prepare data countQC, include=FALSE}
## start of debug
# require(ezRun)
# rawData <- readRDS("/scratch/gtan/p2179-CountQC/Count_QC/rawData.rds")
debug <- FALSE
## end of debug

output <- metadata(rawData)$output
param <- metadata(rawData)$param
seqAnno <- data.frame(rowData(rawData), row.names=rownames(rawData),
                      check.names = FALSE, stringsAsFactors=FALSE)
dataset <- data.frame(colData(rawData), 
                      row.names=colnames(rawData), check.names = FALSE,
                      stringsAsFactors=FALSE)


types = data.frame(row.names=rownames(seqAnno))
for (nm in setdiff(na.omit(unique(seqAnno$type)), "")){
  types[[nm]] = seqAnno$type == nm
}

design = ezDesignFromDataset(dataset, param)
samples = rownames(design)
nSamples = length(samples)
conds = ezConditionsFromDesign(design, maxFactors = 2)
nConds = length(unique(conds))
sampleColors = getSampleColors(conds)
```

```{r check enough samples, echo=FALSE, results='asis'}
if (nSamples < 2){
  cat("Note: Statistics and Plots are not available for single sample experiments.", "\n")
  cat("Run the report again with multiple samples selected.", "\n")
  knit_exit()
}
```

```{r prepare signal, include=FALSE}
signal = shiftZeros(getSignal(rawData), param$minSignal)
presentFlag = assays(rawData)$presentFlag
signalRange = range(signal, na.rm=TRUE)
log2Signal = log2(signal)
isPresent = ezPresentFlags(assays(rawData)$counts, presentFlag=presentFlag, 
                           param=param, isLog=metadata(rawData)$isLog)
signalCond = 2^averageColumns(log2Signal, by=conds)
isPresentCond = averageColumns(isPresent, by=conds) >= 0.5
isPresentStudy <- rowMeans(isPresentCond) >= 0.5
```

## CountQC_Result {.tabset}

### Settings
```{r setting, echo=FALSE}
settings = character()
settings["Normalization method:"] = param$normMethod
if (param$useSigThresh){
  settings["Log2 signal threshold:"] = signif(log2(param$sigThresh), digits=4)
  settings["Linear signal threshold:"] = signif(param$sigThresh, digits=4)
}
settings["Feature level:"] = metadata(rawData)$featureLevel
settings["Number of features:"] = nrow(signal)
settings["Data Column Used:"] = metadata(rawData)$countName
  
kable(settings, row.names=TRUE, 
      col.names="Setting", format="html") %>%
  kable_styling(bootstrap_options = "striped", full_width = F,
                position = "left")
```

```{r output live report, echo=FALSE, results='asis'}
if (exists("output") && !is.null(output)){
  liveReportLink = output$getColumn("Live Report")
  result = EzResult(se=rawData)
  result$saveToFile(basename(output$getColumn("Live Report")))
  cat(paste0("[Live Report and Visualizations](", liveReportLink, ")"), "\n")
}
```

### Data Files

The data files are in tabular text format and can also be opened with a spreadsheet program (e.g. Excel).

When opening with Excel, do make sure that the Gene symbols are loaded into a column formatted as 'text' that prevents
conversion of the symbols to dates). See

(https://www.genenames.org/help/importing-gene-symbol-data-into-excel-correctly)


```{r write data files, include=FALSE}
if(!is.null(assays(rawData)$presentFlag)){
  combined = interleaveMatricesByColumn(assays(rawData)$signal, 
                                        assays(rawData)$presentFlag)
}else{
  combined = assays(rawData)$signal
}
if(!is.null(seqAnno)){
  combined = cbind(seqAnno[rownames(combined), ,drop=FALSE], combined)
}
if(!is.null(combined$featWidth)){
  combined$featWidth = as.integer(combined$featWidth)
}
countFile = paste0(ezValidFilename(param$name), "-raw-count.txt")
ezWrite.table(assays(rawData)$counts, file=countFile, 
              head="Feature ID", digits=NA)
signalFile = paste0(ezValidFilename(param$name), "-normalized-signal.txt")
ezWrite.table(combined, file=signalFile, head="Feature ID", digits=4)

selectSignals = grepl("Signal", colnames(combined))
combined$"Mean signal" = rowMeans(combined[, selectSignals], na.rm=TRUE)
combined$"Maximum signal" = rowMaxs(as.matrix(combined[, selectSignals]),
                                    na.rm = TRUE)
topGenesPerSample = apply(combined[, selectSignals], 2, function(col){
  col = sort(col, decreasing = TRUE)
  if (length(col) > 100) col = col[1:100]
    return(names(col))
  })
    
topGenes = unique(as.character(topGenesPerSample))
    
combined = combined[order(combined$"Maximum signal", decreasing = TRUE), ,
                    drop=FALSE]
useInInteractiveTable = c("seqid", "gene_name", "Maximum signal", 
                          "Mean signal", "description", "featWidth", "gc")
useInInteractiveTable = intersect(useInInteractiveTable, colnames(combined))
tableLink = sub(".txt", "-viewHighExpressionGenes.html", signalFile)
## select top genes
combinedTopGenes = combined[which(rownames(combined) %in% topGenes),]
## restrict number of table rows if necessary
interactiveTable = head(combinedTopGenes[, useInInteractiveTable, drop=FALSE], 
                        param$maxTableRows)
nRows = ifelse(length(topGenes)>=param$maxTableRows, param$maxTableRows,
               length(topGenes))
ezInteractiveTable(interactiveTable, tableLink=tableLink, digits=3, 
                   colNames=c("ID", colnames(interactiveTable)),
                   title=paste("Showing the top", nRows, 
                               "genes with the highest expression"))

rpkmFile = paste0(ezValidFilename(param$name), "-rpkm.txt")
ezWrite.table(getRpkm(rawData), file=rpkmFile, head="Feature ID", digits=4) 
    
tpmFile = paste0(ezValidFilename(param$name), "-tpm.txt")
ezWrite.table(getTpm(rawData), file=tpmFile, head="Feature ID", digits=4)
  
dataFiles = c(countFile, signalFile, rpkmFile, tpmFile, tableLink)

if(isTRUE(param$doZip)){
  dataFiles <- sapply(dataFiles, zipFile)
}
```

```{r add data files link, echo=FALSE, results='asis', message=FALSE}
for(each in dataFiles){
  cat("\n")
  cat(paste0("[", each, "](", each, ")"))
  cat("\n")
}
```

### Count Statistics
```{r count statistics, echo=FALSE, fig.width=min(7+(nSamples-10)*0.3, 30)}
totalCounts = signif(colSums(assays(rawData)$counts) / 1e6, digits=3)
presentCounts = colSums(isPresent)
names(totalCounts) = samples
names(presentCounts) = samples
percentages = paste(signif(100*presentCounts/nrow(isPresent), digits=2), "%")
toPlot <- data.frame(samples=names(totalCounts),
                     totalCounts=totalCounts,
                     presentCounts=presentCounts,
                     percentages=percentages)
m <- list(
  l = 80,
  r = 80,
  b = 200,
  t = 100,
  pad = 0
)
## Total reads
plot_ly(toPlot, x=~samples, 
        y=~totalCounts, type="bar") %>%
  layout(title="Total reads",
         yaxis = list(title = "Counts [Mio]"),
         margin = m
  )
plot_ly(toPlot, x=~samples, y=~presentCounts, type="bar",
        text=~percentages, textposition = 'auto') %>%
  layout(title="Genomic features with reads above threshold",
         yaxis = list(title = "Counts"),
         margin = m
  )
```

### Correlation/Clustering Plot

```{r prepare data correlation, include=FALSE}
assays(rawData)$signal = signal

isValid = isPresentStudy
if (!is.null(seqAnno$IsControl)){
  isValid = isValid & !seqAnno$IsControl
}
  
if(sum(isValid) < 10){
  cat("Not enough valid features for further plots", "\n")
  knit_exit()
}

x = log2(2^log2Signal[isValid, ] + param$backgroundExpression)
xNormed = sweep(x, 1 , rowMeans(x))
xSd <- setNames(rowSds(x, na.rm=TRUE), rownames(x))
ord = order(xSd, decreasing=TRUE)
topGenes = ord[1:min(param$topGeneSize, length(ord))]
```

#### Sample correlation
```{r sample correlation data, include=FALSE}
## All genes
zValues = cor(x, use="complete.obs")
figure.height <- min(max(7, nrow(zValues) * 0.3), 30)
## Top genes
zValuesNormaled = cor(x[topGenes, ], use="complete.obs")
```

```{r plot sample correlation, echo=FALSE, fig.height=figure.height, fig.width=figure.height*1.25}
## All genes
ezCorrelationPlot(zValues, cond=conds, condLabels=conds,
                  main=paste0("all present genes (", sum(isValid), ")"), 
                  colors=sampleColors)
## Top genes
ezCorrelationPlot(zValuesNormaled, cond=conds, condLabels=conds,
                  main=paste0("top genes (", length(topGenes), ")"))
```

```{r sample clustering, echo=FALSE, fig.height=7, fig.width=min(7 + (nSamples-10)*0.3, 30), results='asis'}
if (nSamples > 3){
  cat("#### Sample clustering", "\n")
  ## colour for dendrogram objects
  paletteList <- list("Cavalcanti" = wes_palette("Cavalcanti1"),
                      "Moonrise1" = wes_palette("Moonrise1"),
                      "Darjeeling" = wes_palette("Darjeeling1"),
                      "Royal1" = wes_palette("Royal1"),
                      "FantasticFox" = wes_palette("FantasticFox1"),
                      "Chevalier" = wes_palette("Chevalier1"),
                      'Moonrise2' = wes_palette("Moonrise2"))
  colList = list()
  for (j in 1:ncol(design)) {
    gtab <- unique(design[, j])
    colJ = length(paletteList) - (j %% length(paletteList))
    cols <- colorRampPalette(paletteList[[colJ]])(length(gtab))
    names(cols) = gtab
    colList[[colnames(design)[j]]] = cols
  }
  colAnnot <- design
  for (nm in colnames(colAnnot)){
    colAnnot[,nm] = colList[[nm]][design[,nm]]
  }
  ## All genes
  d = as.dist(1-cor(x, use="complete.obs"))
  hc <- hclust(d, method="ward.D2")
  plotDendroAndColors(hc, colAnnot, autoColorHeight=TRUE, hang = -0.1,
                      main="all present genes")

  ## Top genes
  d = as.dist(1-cor(x[topGenes, ], use="complete.obs"))
  hc <- hclust(d, method="ward.D2")
  plotDendroAndColors(hc, colAnnot, autoColorHeight=TRUE, hang = -0.1,
                      main=paste("top", length(topGenes), " genes"))
}
```

<!--### Correlation/Clustering Plot (Gene-wise)
#### Sample correlation
Gene-wise clustering is not used.
-->
```{r gene wise sample correlation data, include=FALSE, eval=FALSE}
## All genes
zValues = cor(xNormed, use="complete.obs")
figure.height <- min(max(7, nrow(zValues) * 0.3), 30)
## Top genes
zValuesNormaled = cor(xNormed[topGenes,], use="complete.obs")
```

```{r gene wise sample correlation, echo=FALSE, fig.height=figure.height, fig.width=figure.height*1.25, eval=FALSE}
## All genes
ezCorrelationPlot(zValues, cond=conds, condLabels=conds,
                  main=paste0("all present genes (", sum(isValid), ") gene-wise normalized"))
## Top genes
ezCorrelationPlot(zValuesNormaled, cond=conds, condLabels=conds,
                  main=paste0("top genes (", length(topGenes), ") gene-wise normalized"))
```

```{r gene wise sample clustering, echo=FALSE, fig.height=7, fig.width=min(7 + (nSamples-10)*0.3, 30), results='asis', eval=FALSE}
if (nSamples > 3){
  cat("#### Sample clustering", "\n")
  ## All genes
  d <- as.dist(1-cor(xNormed, use="complete.obs"));
  hc <- hclust(d, method="ward.D2")
  #hcd = as.dendrogram(, hang=-0.1)
  #hcd = colorClusterLabels(hcd, sampleColors)
  plotDendroAndColors(hc, colAnnot, autoColorHeight=TRUE, hang = -0.1,
                      main="all present genes; gene-wise normalized")
  
  ## Top genes
  d = as.dist(1-cor(xNormed[topGenes, ], use="complete.obs"));
  hc <- hclust(d, method="ward.D2")
  plotDendroAndColors(hc, colAnnot, autoColorHeight=TRUE, hang = -0.1,
                      main=paste("top", length(topGenes), 
                                 "genes; gene-wise normalized"))
}
```


```{r javascript, echo=FALSE, results='asis', eval=TRUE}
jsFile = system.file("extdata/enrichr.js", package="ezRun", mustWork=TRUE)
invisible(file.copy(from=jsFile, to=basename(jsFile), overwrite=TRUE))
cat(paste0("<SCRIPT language=\"JavaScript\" SRC=\"", basename(jsFile), "\"></SCRIPT>"))
```

```{r Clustering of High Variance Features, echo=FALSE, message=FALSE, fig.width=max(8,4+0.15*ncol(xNormed)), fig.height=10, results='asis', eval=!debug}
clusterPdf = "cluster-heatmap.pdf"
if(nSamples > 3){
  ## gene clustering
  sdThresh = param$highVarThreshold
  use = xSd > sdThresh & !rowAnyMissings(x)
  
  if (sum(use, na.rm=TRUE) > param$maxGenesForClustering){
    use[use] = rank(-1 * xSd[use], ties.method="max") <= param$maxGenesForClustering
    sdThresh = signif(min(xSd[use]), digits=3)
  }
    
  if (sum(use, na.rm=TRUE) > param$minGenesForClustering){
    cat("### Clustering of High Variance Features", "\n")
    cat("\n")
    cat(paste("Threshold for std. dev. of log2 signal across samples:", sdThresh), "\n")
    cat("\n")
    cat(paste("Number of features with high std. dev.:", sum(use)), "\n")
    cat("\n")
    
    #clusterPng = "cluster-heatmap.png"
    clusterColors = c("red", "yellow", "orange", "green", "blue", "cyan")
    clusterResult <-
      clusterPheatmap(xNormed[use, ], design, param, clusterColors,
                      lim=c(-param$logColorRange, param$logColorRange),
                      doClusterColumns=TRUE)
    
    xTmp = data.frame(Cluster=clusterResult$clusterNumbers)
    xTmp$gene_name = seqAnno[rownames(xTmp), "gene_name"]
    xTmp[ ,1] = clusterResult$clusterColors[xTmp[,1]]
    xTmp = xTmp[order(xTmp$Cluster), ,drop=FALSE]
    ezWrite.table(xTmp, file="cluster-heatmap-clusterMembers.txt")

    ## pdf export
    pdf(clusterPdf, width=max(8,4+0.1*ncol(xNormed)), height=10, onefile=FALSE)
    plot(clusterResult$pheatmap$gtable)
    dev.off()
  
    if (doGo(param, seqAnno)){
      clusterResult = goClusterResults(xNormed[use, ], param, clusterResult, 
                                       seqAnno=seqAnno,
                                       universeProbeIds=rownames(seqAnno))
    }
    
    if (!is.null(clusterResult$GO)){
      goTables = goClusterTableRmd(param, clusterResult, seqAnno)
      if (doEnrichr(param)){
        goAndEnrichr = cbind(goTables$linkTable, goTables$enrichrTable)
      } else {
        goAndEnrichr = goTables$linkTable
      }
      bgColors = rep(gsub("FF$", "", clusterResult$clusterColors))
      rownames(goAndEnrichr) <- paste0('<font color="', bgColors,
                                     '">Cluster ', rownames(goAndEnrichr),
                                     '</font>')
    kable(goAndEnrichr, escape=FALSE, row.names=TRUE, format = "html",
          caption="GO categories of feature clusters") %>%
      kable_styling(bootstrap_options = "striped",
                    full_width = F, position = "float_right") %>%
      footnote(general="Cluster font color corresponds to the row colors in the heatmap plot.")
    } else {
      cat("No information available", "\n")
    }
  }
}
```

```{r render go cluster, echo=FALSE, results='hide', message=FALSE, warning=FALSE, eval=!debug}
if (nSamples > 3){
  if (sum(use, na.rm=TRUE) > param$minGenesForClustering){
    if (!is.null(clusterResult$GO)){
    ## GO Cluster tables
    file <- file.path(system.file("templates", package="ezRun"), 
                      "CountQC_goClusterTable.Rmd")
    file.copy(from=file, to=basename(file), overwrite=TRUE)
    rmarkdown::render(input=basename(file), envir = new.env(),
                      output_dir=".", output_file="goClusterTable.html",
                      quiet=TRUE)
    }
  }
}
```

```{r cluster plot link , echo=FALSE, results='asis', eval=!debug}
if(file.exists(clusterPdf)){
    cat(paste0("[Plot in pdf](", clusterPdf, ")"), "\n")
    cat("\n")
}
```

```{r go cluster table link, echo=FALSE, results='asis', eval=!debug}
if (file.exists("goClusterTable.html")){
  cat(paste0("[GO cluster tables](", "goClusterTable.html", ")"), "\n")
  cat("\n")
}
```

```{r cluster heatmap members, echo=FALSE, results="asis", eval=!debug}
if(file.exists("cluster-heatmap-clusterMembers.txt")){
  cat(paste0("[Heatmap cluster members](", "cluster-heatmap-clusterMembers.txt", ")"), 
      "\n")
  cat("\n")
}
```

```{r MDS section, echo=FALSE, results='asis', message=FALSE, eval=TRUE}
if(nSamples > 3){
  cat("### MDS Plot", "\n")
}
```

```{r MDS plot 3D present, echo=FALSE, message=FALSE, fig.height=7, fig.width=9, eval=TRUE}
if(nSamples > 3){
  ## 3D scatter plot is strange. The plots are messy when use htmltools::tagList()
  ## subplot doesn't work either.
  ezMdsPlotly(signal=x, design=design, ndim=3,
              main="mdsPlot_PresentGenes 3D")
}
if(nSamples > 3){
  ezMdsPlotly(signal=x[topGenes,], design=design, ndim=3,
              main="mdsPlot_TopGenes 3D")
}
if(nSamples > 3){
  ezMdsPlotly(signal=x, design=design, ndim=2,
              main="mdsPlot_PresentGenes 2D")
}
if(nSamples > 3){
  ezMdsPlotly(signal=x[topGenes,], design=design, ndim=2,
              main="mdsPlot_TopGenes 2D")
}
```

```{r scatter plot, echo=FALSE, results='asis', eval=!debug}
if(nSamples > 3){
  if (param$writeScatterPlots){
    qcScatterFiles = countQcScatterPlots(param, design, conds, rawData,
                                         signalCond, isPresentCond, types=types)
    if(length(qcScatterFiles) > 0){
      cat("### Scatter Plots by Conditions", "\n")
      for(i in 1:length(qcScatterFiles)){
        cat("\n")
        cat("####", names(qcScatterFiles)[i], "\n")
        for(j in 1:length(qcScatterFiles[[i]])){
          cat("![](",qcScatterFiles[[i]][j],")")
          cat("\n")
        }
        cat("\n")
      }
    }
  }
}
```

```{r density plot, echo=FALSE, results='asis', eval=FALSE}
## Don't run the expression density plot
if(nSamples > 3){
  cat("### Expression densities", "\n")
  cat("Zero or negative counts are not represented by the area!", "\n")
  cat("\n")
  plotCmd = expression({
      #countDensPlot(signal, sampleColors, main="all transcripts", bw=0.7)
      p = countDensGGPlot(cts=data.frame(signal,stringsAsFactors = F),
                          colors=sampleColors, alpha=0.4)
      print(p)
    })
  eval(plotCmd)
}
```

### Input Dataset
```{r input, echo=FALSE}
ezInteractiveTableRmd(values=dataset, digits=4)
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
