---
title: "`r paste('Mothur data analysis report: chimeras, clustering, and phyloseq analysis')`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,warning=FALSE, message=FALSE, out.width = "49%")
## This report requires summary files form mothur
debug <- FALSE
setEnvironments("phantomjs")
```

```{r heatmap, include=FALSE}
### create plots:5. pheatmap
show_pHeatmap <- heatmapForPhylotseqPlotPheatmap(inputData,designMatrix)
```

Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

## Data pre-processing {.tabset}
### Mothur data analysis: summary of the  reads filtered by mapping
```{r MapFiltSummary, echo=FALSE}
kable(finalListOfSummaryTables["MapFiltSummary"]$MapFiltSummary$mergedTable, escape=TRUE, row.names=FALSE, format="html") %>%
  kable_styling(bootstrap_options = c("striped", "bordered"),
                full_width = T) %>%
  group_rows(index = finalListOfSummaryTables["MapFiltSummary"]$MapFiltSummary$aboveHeader) %>%
  scroll_box(width = "100%", height = "600px")
```

### Mothur data analysis: summary of the reads preclustered and filtered for chimeras
```{r,echo=FALSE, out.width="60%"}
  plot(chimeraPlotsToReport)
```

### Mothur data analysis: OTUs convergence summary
```{r stepConvergence, echo=FALSE}
kable(convStepTableToReport["stepConvergenceSummary"]$stepConvergenceSummary$mergedTable, escape=TRUE, row.names=FALSE, format="html") %>%
  kable_styling(bootstrap_options = c("striped", "bordered"),
                full_width = T) %>%
  group_rows(index = convStepTableToReport["stepConvergenceSummary"]$stepConvergenceSummary$aboveHeader) %>%
  scroll_box(width = "100%", height = "600px")
```

### Mothur data analysis: OTUs saturation plots
```{r,echo=FALSE, out.width="60%"}
  plot(otuSaturPlotToReport)
```

### Phyloseq analysis: abundance barplots
#### If samples are grouped, the second plot is more informative
```{r abund, echo=FALSE}
par(mfrow=c(1,2), las=1)
plot(abundPlot)
plot(abundPlotGroup)
```

### Phyloseq analysis: clustering (ordination) plots
```{r ordinTaxa, echo=FALSE,eval=isGroupThere}
par(mfrow=c(1,2), las=1)
plot(plotOrdTaxa)
plot(plotOrdSamples)
```

### Phyloseq analysis: richness plots
```{r rich, echo=FALSE,out.width="60%"}
plot(plotRichness)
```

```{r tree, echo=FALSE,out.width="100%", eval=FALSE}
plot(plotTree)
```

### Phyloseq analysis:rarefaction plots
```{r raref, echo=FALSE}
par(mfrow=c(1,2), las=1)
for (RarPlot in rarefPlot){
plot(RarPlot)
}
```

### Heatmaps
```{r heat, echo=FALSE,out.width="100%"}
show_pHeatmap()
```

### Phyloseq analysis: two-goups comparison
```{r volcano, echo=FALSE,eval=isGroupThere, out.width="33%"}
par(mfrow=c(1,3), las=1)
plot(deseqResults$vPlot)
plot(deseqResults$logPlot)
plot(deseqResults$pieChart)
```

#### DESeq analysis: summary table (top 20 differentially expressed organisms)
```{r deSeqTable, echo=FALSE,eval=isGroupThere,out.width="100%"}
kable(deseqResults$table, escape=TRUE, row.names=FALSE, format="html") %>%
  kable_styling(bootstrap_options = c("striped", "bordered"),
                full_width = T) %>%
  scroll_box(width = "100%", height = "600px")
```

### SessionInfo
```{r, echo=FALSE}
sessionInfo()
```
