---
title: "`r paste('BAM Statistics:', param$name)`"
author: "Functional Genomics Center Zurich"
output: 
  html_document:
    self_contained: false
    includes:
      in_header: fgcz_header.html
    css: fgcz.css
---
Started on `r format(Sys.time(), "%Y-%m-%d %H:%M:%S")`

```{r setup RNABamStats, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
## This report requires the object resultList, dataset, param
require(knitr)
require(kableExtra)
```

```{r prepare data RNABamStats, include=FALSE}
## debug:
#require(ezRun)
#load("/home/gtan/analysis/gtan/p2438-RNABamStats/resultList.RData")
#load("/home/gtan/analysis/gtan/p2438-RNABamStats/dataParam.rda")
## end of debug

conds = ezConditionsFromDataset(dataset, param=param)
samples = rownames(dataset)
sampleColors = getSampleColors(conds, samples)
bamFiles = dataset$BAM
```

```{r igv link, include=FALSE}
if (param$writeIgvSessionLink){
#     titles[["Genome Browser"]] = "Genome Browser"
#     addTitle(doc, titles[[length(titles)]], 2, id=titles[[length(titles)]]) ## NOTEP: igv links are not yet added to the report.
  if (length(bamFiles) > 4){
    idx = which(!duplicated(conds))
    idx = idx[1:min(4, length(idx))]
  } else {
    idx = 1:length(bamFiles)
  }
  for (each in idx){
    writeIgvSession(genome=getIgvGenome(param), refBuild=param$ezRef["refBuild"],
                    file=basename(sub(".bam", "-igv.xml", bamFiles[each])),
                    bamUrls=paste(PROJECT_BASE_URL, bamFiles[each], sep="/"))
    writeIgvJnlp(jnlpFile=basename(sub(".bam", "-igv.jnlp", bamFiles[each])),
                 projectId=param$projectId,
                 sessionUrl=paste(PROJECT_BASE_URL, sub(".bam", "-igv.xml", 
                                                        bamFiles[each]), 
                                  sep="/"))
  }
}
```

## RNA_BAM_Statistics {.tabset}

<!-- ### Read Alignment Statistics -->

### Multi-Matching Reported in Bam File
The plot holds for each sample the number of reads in Millions 
that have X matches in the target and are reported in the file.

```{r multi-matching data, echo=FALSE}
mmValues = integer()
for (sm in samples){
  mmValues = union(mmValues, as.integer(names(resultList[[sm]]$multiMatchInFileTable)))
}
mmCounts = ezMatrix(0, rows=samples, cols=sort(mmValues))
for (sm in samples){
  mm = resultList[[sm]]$multiMatchInFileTable
  mmCounts[sm, names(mm)] = mm
}
```

```{r multi-matching plot, echo=FALSE, message=FALSE, fig.height=7, fig.width=min(max(7, 7+(nrow(mmCounts)-20)*0.3), 30)}
alignmentCountBarPlot(mmCounts, relative=FALSE,
                      file="multiMatchInFile-barplot.txt")
alignmentCountBarPlot(mmCounts, relative=TRUE)
```

```{r multi-matching export, echo=FALSE}
txtFile <- "read-alignment-statistics.txt"
ezWrite.table(mmCounts, file=txtFile, head="Sample")
```
[`r txtFile`](`r txtFile`)

```{r MatchTargetTypeCounts, echo=FALSE, results='asis', message=FALSE, fig.width=8, fig.height=7, warning=FALSE}
for (nm in c("multiMatchTargetTypeCounts", "uniqueMatchTargetTypeCounts")){
    if (!is.null(resultList[[1]][[nm]])){
      readSet = switch(nm,
                       multiMatchTargetTypeCounts="Uniquely and multi-matching reads:",
                       uniqueMatchTargetTypeCounts="Uniquely matching reads:")
      cat("###", paste(readSet, "Match Count Percentages"), "\n")
      tct = getTypeCountTable(resultList, nm)
      ezWrite.table(tct, file=paste0(nm, ".txt"), digits=4)
      tpt = as.matrix(tct)
      for (cn in colnames(tpt)){
        tpt[ ,cn] = tct[ ,cn]/ tct["total", cn] * 100
      }
      minPercentage = 1
      rowsUse = setdiff(rownames(tpt)[apply(tpt, 1, max) > minPercentage], "total")
      tptUse = tpt[rowsUse, , drop=FALSE]
      if (nrow(tptUse) >= 2 && ncol(tptUse) >= 2){
        tptUseRel = log2(tptUse)
        meanVals = apply(tptUseRel, 1, function(x){mean(x[is.finite(x)])}) ## ignore zero counts
        tptUseRel = tptUseRel - meanVals
        plotCmd = expression({
          ezHeatmap(tptUseRel, margins=c(10, 12), lim=c(-2, 2),
                    Rowv=FALSE, Colv=FALSE, main="Relative Prevalence [log2]")
        })
        eval(plotCmd)
        k <- kable(signif(tptUse, digits=3),
              row.names=TRUE, format = "html",
              caption="Match Count Percentages") %>%
          kable_styling(bootstrap_options = "striped",
                        full_width = F, position = "left") %>%
          footnote(general="Percentage value.")
        print(k)
      }else{
        cat("The plot and table are shown when there are at least two samples!",
            "\n")
      }
      
      cat("###", paste(readSet, "Read Starts per Base"), "\n")
      tct = as.matrix(getTypeCoverageTable(resultList, nm))
      ezWrite.table(tct, file=paste0(nm, "-coverage.txt"), digits=4)
      if (nrow(tct) >= 2 && ncol(tct) >= 2){
        tctRel = log2(sweep(tct, 2, tct["total", ], FUN="/"))
        tctRel = tctRel[rowsUse, , drop=FALSE]
        plotCmd = expression({
          ezHeatmap(tctRel, margins=c(10, 12), lim=c(-5, 5),
                    Rowv=FALSE, Colv=FALSE, main="Coverage Enrichment")
        })
        eval(plotCmd)
        k <- kable(signif(tct[rowsUse, ], digits=4),
                   row.names=TRUE, format = "html",
                   caption="Read Starts per Base") %>%
          kable_styling(bootstrap_options = "striped",
                        full_width = F, position = "left") %>%
          footnote(general="Read Starts per Base is equivalent to Coverage divided by Read length.")
        print(k)
      }else{
        cat("The plot and table are shown when there are at least two samples!",
            "\n")
      }
    }
}

```

```{r transcript coverage, echo=FALSE, results='asis'}
if (!is.null(resultList[[1]][["TranscriptsCovered"]])){
  cat("###", "Coverage plot\n")
  cat("####", "Transcripts covered plot\n")
  minYlim = 0 #min(sapply(resultList, function(item){min(item[["TranscriptsCovered"]][["counts"]])}))
  minXlim = 0
  maxYlim = max(sapply(resultList,
                  function(item){max(item[["TranscriptsCovered"]][["counts"]])}))
  maxXlim = 130 #max(sapply(resultList, function(item){max(item[["TranscriptsCovered"]][["mids"]])}))
    
  plotCmd = expression({
    plot(1, 1, xlim=c(minXlim, maxXlim), ylim=c(minYlim, maxYlim), xlab = "% length covered", ylab="number of isoforms", main="Isoforms Covered Histogram", type="n")
      for (sm in samples){
        transcript_covered = resultList[[sm]][["TranscriptsCovered"]]
        lines(transcript_covered[["mids"]], transcript_covered[["counts"]], col=sampleColors[sm])
      }
      legend("topright", samples, col=sampleColors[samples], cex=1.2, pt.cex=1.5, bty="o", pt.bg="white", lty=1)
    })
  eval(plotCmd)
}
```

```{r genebody coverage legend, echo=FALSE, fig.width=3, fig.height=0.25*(length(samples)+1L), results='asis'}
if (length(resultList[[1]][["genebody_coverage"]]) != 0){ ## TODO this could be done better by searching alls results for a valid genebody_coverage element
  cat("####", "Genebody coverage plot\n")
  plotCmd = expression({
    ezLegend(legend=samples, fill=sampleColors[samples], title="Sample Colors")
    })
  eval(plotCmd)
}
```

```{r genebody coverage, echo=FALSE, results='asis'}
if (length(resultList[[1]][["genebody_coverage"]]) != 0){ ## TODO this could be done better by searching alls results for a valid genebody_coverage element
    minYlim = 0 #min(sapply(resultList, function(item){min(item[["genebody_coverage"]])}))
    maxYlim = 0.08 ## this means we allow at most 10-fold enrichment at a percentile #max(sapply(resultList, function(item){max(item[["genebody_coverage"]])}))
    gbcTemplate = resultList[[1]][["genebody_coverage"]]
    pngMatrix = ezMatrix("", rows=names(gbcTemplate),
                         cols=names(gbcTemplate[[1]]))
    #pngLinks = character()
    for (rn in c("less than 400nt", "1000 to 1200nt")){ # rownames(pngMatrix)){
      for (cn in "medium expressed"){ #} colnames(pngMatrix)){
        pngMatrix[rn, cn] = ezValidFilename(paste0("genebody_coverage_", rn, "_", cn, ".png"))
        covValues = ezMatrix(0, cols=0:100, rows=samples)
        for (sm in samples){
          y = resultList[[sm]][["genebody_coverage"]][[rn]][[cn]]
          if (!is.null(y)){
            covValues[sm, ] = y
          } else {
            covValues[sm, ] = NA
          }
        }
        plotCmd = expression({
          plot(1, 1, xlim=c(0,100), ylim=c(minYlim, maxYlim), xlab="percentile of geneBody (5'->3')", ylab="relative coverage", 
               main=paste("Genebody coverage", rn, cn), type="n",
               axes=FALSE, frame=TRUE)
          axis(side=2)
          axis(side=1, at=seq(0, 100, by=10))
          for (sm in samples){
            y = resultList[[sm]][["genebody_coverage"]][[rn]][[cn]]
            if (!is.null(y)){
              covValues[sm, ] = y
              lines(0:100, y, col=sampleColors[sm])
            } else {
              covValues[sm, ] = NA
            }
          }
        })
        eval(plotCmd)
       
        ezWrite.table(covValues, file=sub(".png$", ".txt", pngMatrix[rn, cn]),
                      head="Name")
      }
    }
  }
```

```{r countQC, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
geneCounts = resultList[[1]][["geneCounts"]]
if (!is.null(geneCounts) && !is.null(param$skipCountQc) && !param$skipCountQc){
    counts = ezMatrix(0, rows=names(geneCounts), cols=names(resultList))
    for (i in 1:length(resultList)){
      counts[ ,i] = resultList[[i]][["geneCounts"]][rownames(counts)]
    }
    seqAnno = ezFeatureAnnotation(param, rownames(counts), "gene")
    require(SummarizedExperiment)
    rawData <- SummarizedExperiment(
      assays=SimpleList(counts=counts, presentFlag=counts > param$sigThresh),
      rowData=seqAnno, colData=dataset,
      metadata=list(isLog=FALSE, featureLevel="gene",
                    type="Counts", countName="multiMatchCounts",
                    param=param)
    )
    if (ezIsSpecified(param$transcriptTypes)){
      use = seqAnno$type %in% param$transcriptTypes
    } else {
      use = TRUE
    }
    rawData <- rawData[use, ]
    assays(rawData)$rpkm = getRpkm(rawData)
    assays(rawData)$tpm = getTpm(rawData)
    
    if (is.null(param$normMethod)){
      param$normMethod = "logMean"
    }
    if (!ezIsSpecified(param$runGO)){
      param$runGO = TRUE
    }
    metadata(rawData)$param <- param
    setwdNew("Count_QC")
    assays(rawData)$signal = ezNorm(assays(rawData)$counts,
                                    presentFlag=assays(rawData)$presentFlag,
                                    method=param$normMethod)
    styleFiles <- file.path(system.file("templates", package="ezRun"),
                          c("fgcz.css", "CountQC.Rmd",
                            "fgcz_header.html", "banner.png"))
    file.copy(from=styleFiles, to=".", overwrite=TRUE)
    output <- NULL ## No live report 
    rmarkdown::render(input="CountQC.Rmd", envir = new.env(),
                      output_dir=".", output_file=htmlFile, quiet=TRUE)
    setwd("..")
    cat("[Count QC Report](Count_QC/00index.html)\n")
}
```

```{r junction saturation plot for all samples, echo=FALSE, results='asis'}
if(!is.null(resultList[[1]][["Junction"]])){
  cat("###", "Junction saturation plot for all samples\n")
  junctionMaxVal = numeric()
  for (nm in names(resultList[[1]][["Junction"]])){
    junctionMaxVal[nm] = 0
    for (sm in samples){
      junctionMaxVal[nm] = max(junctionMaxVal[nm], unlist(resultList[[sm]][["Junction"]][[nm]]))
    }
  }
  for (nm in names(resultList[[1]][["Junction"]][["junctionSaturation"]])){
    plotCmd = expression({
        plot(1,1, xlab="percent of total reads", 
             ylab='Number of splicing junctions (x1000)',
             type='o', xlim=c(0, 130),
             ylim=c(0, junctionMaxVal["junctionSaturation"]/1000), main=nm)
        for (sm in samples){
          x <- resultList[[sm]][["Junction"]][["junctionSaturation"]][[nm]]
          lines(as.numeric(names(x))*100, x/1000, col=sampleColors[sm], type="o")
        }
        legend("bottomright", legend=samples, col=sampleColors[samples], 
               lwd=1, pch=1)
      })
    eval(plotCmd)
  }
}
```

```{r junction plot, echo=FALSE, results='asis'}
if (!is.null(resultList[[1]][["Junction"]])){
  cat("###", "Junction plots\n")
  for (sm in samples){
    #pngLinks = character()
    cat("\n")
    cat("####", sm, "\n")
    for (nm in names(resultList[[sm]][["Junction"]])){
        junctionPlot = resultList[[sm]][["Junction"]][[nm]]
        plotCmd = expression({
          if (nm %in% c("splice_events", "splice_junction")){
            pie(junctionPlot, col=c(2,3,4), init.angle=30, angle=c(60,120,150),
                density=c(70,70,70),main=nm, 
                labels=paste(names(junctionPlot), paste0(round(junctionPlot), "%")))
          } else if (nm =="junctionSaturation"){
            x = as.numeric(names(junctionPlot[[1]])) * 100
            plot(1,1,xlab="percent of total reads", ylab='Number of splicing junctions (x1000)',type='o',
                 ylim=c(0, junctionMaxVal[nm]/1000), xlim=range(x))
            saturationColors = c("all junctions"="blue", "known junctions"="red", "novel junctions"="green")
            for (item in names(junctionPlot)){
              lines(x, junctionPlot[[item]]/1000, col=saturationColors[item], type="o")
            }
            legend("topleft", legend=names(saturationColors), col=saturationColors,lwd=1,pch=1)
          }
        })
        eval(plotCmd)
    }
    cat("\n")
  }
}
```

```{r fragSizeHist, echo=FALSE, results='asis'}
if(!is.null(resultList[[1]]$fragSizeHist)){
  cat("###", "Length distribution of fragments for paired reads\n")
  for (sm in samples){
    fsh = resultList[[sm]]$fragSizeHist
    #pngFiles[sm] = paste0(sm, "-fragSizeHist.png")
    ezWrite.table(cbind(Length=fsh$mids, Count=fsh$counts),
                  file=paste0(sm, "-fragSizeHist.txt"), row.names=FALSE)
    plotCmd = expression({
        plot(fsh, xlab="fragment size", main=paste(sm, "-- Length Histogram"), 
             ylim=c(0, max(fsh$counts[-length(fsh$counts)]))) ## don't use the longest fragment size
    })
    try({
      eval(plotCmd)
      #pngFiles[sm] = ezImageFileLink(plotCmd, file=pngFiles[sm], width=600)
    })
  }
}
```

### Histogram of aligned segments per read
```{r Histogram of aligned segments per read, echo=FALSE, results='asis'}
for (sm in samples){
  fsh = resultList[[sm]]$segmentCountHist
  #pngFiles[sm] = paste0(sm, "-segmentCountHist.png")
  ezWrite.table(cbind(Length=fsh$mids, Count=fsh$counts),
                file=paste0(sm, "-segmentCountHist.txt"), row.names=FALSE)
    
  plotCmd = expression({
    plot(fsh, xlab="# segments in alignment", 
         main=paste(sm, "-- Histogram of Segments per Alignment"))
  })
  try({
    #pngFiles[sm] = ezImageFileLink(plotCmd, file=pngFiles[sm], width=600)
    eval(plotCmd)
  })
}
```

```{r Read position specific error rate, echo=FALSE, results='asis', fig.width=16}
if (!is.null(resultList[[1]][["ErrorRates"]])){
  cat("###", "Read position specific error rate\n")
  for (sm in samples){
    for (nm in names(resultList[[sm]][["ErrorRates"]])){
      errorRate = resultList[[sm]][["ErrorRates"]][[nm]]
      if (!is.null(errorRate)){
        pngFile = ezValidFilename(paste0(sm, "_", nm, ".png"))
        plotCmd = expression({
          plotPosSpecificErrorRate(errorRate, png=pngFile, main=paste(sm, nm))
        })
        eval(plotCmd)
        #pngLinks[nm] = ezImageFileLink(plotCmd, file=pngFile, width=1600)
      }
    }
  }
}
```

```{r The fraction of isoform length covered, echo=FALSE, results='asis'}
if (!is.null(resultList[[1]][["TranscriptsCovered"]])){
  cat("###", "The fraction of isoform length covered\n")
  for (sm in samples){
    tlc = resultList[[sm]][["TranscriptsCovered"]]
    #pngFiles[sm] = paste0(sm, "-transcriptsCovered.png")
    ezWrite.table(cbind(Percents=tlc$mids, Count=tlc$counts),
                  file=paste0(sm, "-transcriptsCovered.txt"), row.names=FALSE)
    plotCmd = expression({
      cts = tlc$counts
      names(cts) = tlc$mids
      barplot(cts, xlab="% length covered", main=paste(sm, "-- Isoforms Covered Histogram"), ylab="number of isoforms")
    })
    #pngLinks[sm] = ezImageFileLink(plotCmd, file=pngFiles[sm], width=600)
    eval(plotCmd)
    #plotBySamples[[sm]] = c(plotBySamples[[sm]], pngLinks[sm])
    #plotByStatistics[["The fraction of isoform length covered"]] = c(plotByStatistics[["The fraction of isoform length covered"]], pngLinks[sm])
  }
}
```

```{r duplication rates, echo=FALSE, results='asis'}
if(!is.null(resultList[[1]][["dupRate"]])){
  cat("###", "Duplication rate quality control\n")
  cat("\n")
  cat("The number of reads per base assigned to a gene in an ideal RNA-Seq data set is expected to be proportional to the abundance of its transcripts in the sample. For lowly expressed genes we expect read duplication to happen rarely by chance, while for highly expressed genes - depending on the total sequencing depth - we expect read duplication to happen often.", "\n")
  cat("\n")
  require(dupRadar, quietly = TRUE)
  for(sm in samples){
    duprateExpDensPlot(DupMat=resultList[[sm]][["dupRate"]])
    title(paste(sm, "-- 2D density scatter plot"))
  }
}
```

### Input Dataset
```{r, echo=FALSE, message=FALSE}
ezInteractiveTableRmd(dataset)
```